tests/fsharpqa/Source/nunitConf.fs

```diff
@@ -0,0 +1,420 @@
+
+
+[<assembly:ParallelizableAttribute(ParallelScope.Fixtures)>]
+[<assembly:InitializeSuite()>]
+()
+
+module FSharpTestSuite =
+
+    let getTagsOfFile path =
+        match File.ReadLines(path) |> Seq.truncate 5 |> Seq.tryFind (fun s -> s.StartsWith("// #")) with
+        | None -> []
+        | Some line -> 
+            line.TrimStart('/').Split([| '#' |], StringSplitOptions.RemoveEmptyEntries)
+            |> Seq.map (fun s -> s.Trim())
+            |> Seq.filter (fun s -> s.Length > 0)
+            |> Seq.distinct
+            |> Seq.toList
+
+    let getTestFileMetadata dir =
+        Directory.EnumerateFiles(dir, "*.fs*")
+        |> Seq.toList
+        |> List.collect getTagsOfFile
+
+    let parseTestLst path =
+        let dir = Path.GetDirectoryName(path)
+        let commentLine (t: string) = t.StartsWith("#")
+        let lines =
+            File.ReadAllLines(path)
+            |> Array.filter (not << commentLine)
+            |> Array.filter (not << String.IsNullOrWhiteSpace)
+        let parse (t: string) =
+            let a = t.Split([| '\t'; '\t' |], StringSplitOptions.RemoveEmptyEntries)
+            let testDir = Commands.getfullpath dir a.[1]
+            [| for x in a.[0].Split(',') do yield (x, testDir) |]
+
+        lines |> Array.collect parse |> List.ofArray
+
+    let ``test.lst`` = lazy ( 
+        parseTestLst ( __SOURCE_DIRECTORY__/"test.lst" ) 
+        )
+
+    let getTestLstTags db dir =
+        let normalizePath path =
+            Uri(path).LocalPath
+            |> (fun s -> s.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar))
+            |> (fun s -> s.ToUpperInvariant())
+
+        let sameDir a = (normalizePath dir) = (normalizePath a)
+        let normalizedPath = normalizePath dir
+        db
+        |> List.choose (fun (tag, d) -> if sameDir d then Some tag else None)
+
+    let fsharpSuiteDirectory = __SOURCE_DIRECTORY__
+
+    let setProps dir (props: NUnit.Framework.Interfaces.IPropertyBag) =
+        let testDir = dir |> Commands.getfullpath fsharpSuiteDirectory
+
+        if not (Directory.Exists(testDir)) then failwithf "test directory '%s' does not exists" testDir
+
+        let categories = [ dir ] @ (testDir |> getTestFileMetadata) @ (testDir |> getTestLstTags ``test.lst``.Value)
+        categories |> List.iter (fun (c: string) -> props.Add(NUnit.Framework.Internal.PropertyNames.Category, c))
+
+        props.Set("DIRECTORY", testDir)
+
+    let testContext () =
+        let test = NUnit.Framework.TestContext.CurrentContext.Test
+        { Directory = test.Properties.Get("DIRECTORY") :?> string;
+          Config = suiteHelpers.Value }
+
+// parametrized test cases does not inherits properties of test ( see https://github.com/nunit/nunit/issues/548 )
+// and properties is where the custom context data is saved
+
+type FSharpSuiteTestAttribute(dir: string) =
+    inherit NUnitAttribute()
+    interface NUnit.Framework.Interfaces.IApplyToTest with
```

Comment:
These more advanced uses of nunit are a bit obscure for those of us used to using nunit in "simple" mode. Could you add a more explanation as to what's going on please? Thanks